Source/DafnyCore/AST/Formatting.dfy

```diff
@@ -0,0 +1,176 @@
+      }
+    }
+  }
+  class {:extern "HelperString"} {:compile false} HelperString {
+    static predicate FinishesByNewline(input: CsString)
+  }
+}
+module {:extern "Microsoft"} {:options "-functionSyntax:4"}  Microsoft {
+  module {:extern "Dafny"} Dafny {
+    module {:extern "Formatting"} Formatting {
+      import opened MicrosoftDafny
+      import opened System
+
+      const {:extern "System", "String.Empty"} CsStringEmpty: CsString;
+
+      trait IIndentationFormatter {
+        function Reindent(token: IToken, trailingTrivia: bool, precededByNewline: bool, indentation: CsString, lastIndentation: CsString): CsString
+        method GetIndentation(token: IToken, currentIndentation: CsString)
+          returns (
+            indentationBefore: CsString,
+            indentationBeforeSet: bool,
+            lastIndentation: CsString,
+            lastIndentationSet: bool,
+            indentationAfter: CsString,
+            indentationAfterSet: bool)
+          requires token.Valid()
+          ensures token.allTokens == old(token.allTokens)
+      }
+
+      lemma IsAllocated<T>(x: T)
+        ensures allocated(x) {
+      }
+
+      /** Prints the entire program while fixing identation, based on
+          1) indentation information provided by the IIndentationFormatter reindent
+          2) Reindentation algorithm provided by the same reindent */
+      method printSourceReindent(firstToken: IToken, reindent: IIndentationFormatter) returns (s: CsString)
+        requires firstToken.Valid()
+        ensures forall token <- firstToken.allTokens :: s.Contains(token.val)
+      {
+        var token: IToken? := firstToken;
+        var sb := new CsStringBuilder();
+        var currentIndent := CsStringEmpty;
+        ghost var i := 0;
+        var leadingTriviaWasPreceededByNewline := true;
+        var allTokens := firstToken.allTokens;
+        while(token != null)
+          decreases if token == null then 0 else |token.allTokens|
+          invariant 0 <= i <= |allTokens|
+          invariant if token == null then i == |allTokens| else
+                    && token.Valid()
+                    && i < |allTokens|
+                    && token == allTokens[i]
+          invariant forall t <- allTokens[0..i] :: sb.built.Contains(t.val)
+        {
+          if(token.Next == null) {
+            firstToken.TokenNextIsNullImpliesLastToken(token, i);
+          } else {
+            firstToken.TokenNextIsIPlus1(token, i);
+          }
+          IsAllocated(allTokens[0..i]);
+          var indentationBefore, indentationBeforeSet,
+              lastIndentation, lastIndentationSet,
+              indentationAfter, indentationAfterSet := reindent.GetIndentation(token, currentIndent);
+
+          var newLeadingTrivia := reindent.Reindent(token, false, leadingTriviaWasPreceededByNewline, indentationBefore, lastIndentation);
+          var newTrailingTrivia := reindent.Reindent(token, true, false, indentationAfter, indentationAfter);
```

Comment:
What is there to reindent for trailing trivia? How can you reindent something that's not at the start of the line? Can trailing trivia be at the start of a line?

Reply:
Trailing trivia can have newlines. Trailing trivia stops after one blank newline.